### DISCUSSION TITLE
Where to define environment based CF variables

### DISCUSSION THREAD

**Discussion by cfcoder** [wheels-dev/wheels#1347] [Q&A] [answered]
I want to set some variables that differ for development environment vs production environment. I notice in the settings.cfm in /config, values for an environment use the set() function, but that just sets values into CFWheels configurations. I'm looking for a place to set  global variables that are dependent on the environment I'm running in. I'd also like to know where to set global variables that don't have anything to do with the environment, like a lookup struct of values that don't change and I want them defined and accessible everywhere.

**Comment by bpamiri** [comment]
@cfcoder Bruce you can set those in the various config files. You can put any code you want in those files in addition to framework settings.

For instance put all common globals in the `config > settings.cfm` file or set environment based globals in each environment specific file like `config > development > settings.cfm` or `config > production > settings.cfm`.

**Reply by cfcoder** [reply]
I tried it there but it didn't work. But I think I failed to reload.

What about using the /views/helpers.cfm? I just found it and noticed it is for functions with global access. It appears variables defined there are also global. What is best practice?

**Reply by cfcoder** [reply]
Setting variables in /config/development/settings.cfm isn't working for me, even after reload. The variable is not recognied in the layout.cfm. Gives error that it is undefined. Do I need to use the set function?

**Reply by bpamiri** [reply]
The /views/helpers.cfm file is intended for custom functions that are available to your views files. There is no init life cycle on it like the config files would have. But you could tcheorerially have code that interrogates the environment variable and performs one function or the other but that's not the intended use case. I think you'd be better served going with the config files. Keep in mind that settings in the config files and models are cached so you have to do a reload for changes to be picked up.

**Reply by cfcoder** [reply]
I'm not having success with a simple 
foobar=42;
line placed in config/development/settings.cfm file and then checking it's value in layout.cfm.  I reloaded and foobar is still undefined. I even restarted my CF server and it still comes up undefined.

**Reply by bpamiri** [reply]
You'll have to define your globals in a more "global" scope. Try session.foobar or application.foobar.

**Reply by cfcoder** [reply]
So there is no designated place to add global variables without using the session or application (or client for that matter) scopes?

**Reply by bpamiri** [reply]
That's correct since at the end of the day you are in a CFML environment. The exception to this is variables created in the controller are available to the corresponding view file. 

This really depends on what you are trying to do. If you are trying to define lookup data that isn't in a database then a table-less model may work for you.

If you are trying to create some environmental settings that you want to keep out of the code base and source control you may be able to use a .env file. 

It really depends on how you intend to use the data.

**Reply by cfcoder** [reply]
Thanks for your suggestions. Since my environmental need was for some auto-logout settings, session scope will work good. I have a session.cfc controller so I set the values in the session scope in that file, based on the get("environment") value.

**Reply by cfcoder** [reply]
@bpamiri Doesn't it work to set any shared variables in a controller's config() function? I tried it, but it doesn't seem to work. It's not clear what you can set in the config function of each controller and when it gets run.

**Reply by neokoenig** [reply, marked as answer]
The `config()` definition is cached, so if you set `request.foo = 1` within `config()` directly, it will only be run once, when the controller is initialised (so only a `?reload=true` / application restart will re-examine it).

What you can do is use a filter to setup things, which, whilst the config() function is cached, will get run.

i.e, 

```
function config(){
  filters(through="setDefaults");
}

private function setDefaults(){
  // I will get executed on each request
  request.foo = 1;
}
```

**Comment by adamcameron** [comment]
> session.cfc controller so I set the values in the session scope in that file

It's not the job of a controller to be doing stuff like this. A controller should simply be marshalling incoming request values, calling the model layer to do [whatever data processing is needed for the given request], and ensuring the view gets the data it needs. This sort of thing is the job of something in the model tier.

Stuff relating to sessions should be _set_ in `/events/onsessionstart.cfm`
Env-specific values should be handled in `/config/[env]/settings.cfm`

In `settings.cfm` I would init an Environment model with whatever the values are, and when needing to do stuff like set a session variable with a value from that environment, fetch it back out from the Environment model (this fetch / assign being in `onsessionstart.cfm`).
